### PR TITLE
Rest APIs for Biometric Device Registration

### PR DESCRIPTION
**Pull request by avishkajayasundara**

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            <artifactId>org.wso2.carbon.identity.api.user.common</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.outbound.auth.biometric</groupId>
+            <artifactId>org.wso2.carbon.identity.application.authenticator.biometric.device.handler</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/pom.xml]:
```suggestion
</project>

```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+package org.wso2.carbon.identity.api.user.biometric.device.common.util;
+
+/**
+ * Constants used in Biometric Device handler Rest APIs.
+ */
+public class Constants {
+
+      public enum ErrorMessages {
+        ERROR_CODE_INVALID_SIGNATURE(
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]:
```suggestion
      public enum ErrorMessages {
      
        ERROR_CODE_INVALID_SIGNATURE(
```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                "BDH-15001",
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]: What does BDH stand for? :)

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                "The credibility of the request could not be verified" ,
+                "An error occurred when the digital signatre is not verified"),
+
+        ERROR_CODE_DEVICE_HANDLER_SQL_EXCEPTION(
+                "BDH-15002",
+                "The operation was not executed due to a database error",
+                "A SQL exception occurred while trying to perform database operation "),
+        ERROR_CODE__JSON_PROCESSING_EXCEPTION(
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]: Can we recheck whether code formatting is applied?

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                "BDH-15003",
+                "Error occured while trying to convert an object to json format",
+                "" + "A System error occurred while trying to convert an object to a json"),
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]:
```suggestion
                "A System error occurred while trying to convert an object to a JSON"),
```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        ERROR_CODE_EDIT_DEVICE_NAME_CLIENT_ERROR(
+                "BDH-10005",
+                "The device could not be modified",
+                "A System error occurred while trying to modify the device name");
+
+        private final String code;
+        private  String message;
+        private final String description;
+
+          ErrorMessages(String code, String message, String description) {
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]:
```suggestion
         ErrorMessages(String code, String message, String description) {
```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+        public String getDescription() {
+            return description;
+        }
+
+        @Override
+        public String toString() {
+            return getCode() + " | " + getMessage() + " | " + getDescription();
+        }
+
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]:
```suggestion
```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/HandleException.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import java.security.NoSuchAlgorithmException;
+import java.security.SignatureException;
+import java.sql.SQLException;
+import javax.mail.AuthenticationFailedException;
+import javax.ws.rs.core.Response;
+
+/**
+ * The class which handles API errors.
+ */
+public class HandleException {
```
Comment (tharindu-bandara) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/HandleException.java]: Check code formatting.

Also, what is the reason for separating utilities into specific classes? Can we use a single util class?

For example, calling the static method  `handleException` needs to be done in the following manner.
```
HandleException.handleException(...)
```

Having a single util class would improve the readability, and separation of components.
```
BiometricDeviceUtils.handleException(...)
```
```

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]: Year is wrong, have you updated the licence template?

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.handler.v1/.openapi-generator/biometricdevice.yaml.sha256 [hunk 10]
```diff
@@ -0,0 +1 @@
+295d9adebc69e33fdaccff538d476a5b4439946250e137f871fcd17bde0cebac
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.handler.v1/.openapi-generator/biometricdevice.yaml.sha256]: is this a local auto-generated file?

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.handler.v1/pom.xml [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+        <dependency>
+            <groupId>org.apache.cxf</groupId>
+            <artifactId>cxf-rt-frontend-jaxrs</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.cxf</groupId>
+            <artifactId>cxf-rt-frontend-jaxws</artifactId>
+            <version>3.3.5</version>
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.handler.v1/pom.xml]: Move the version to parent's dependency management

File: components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java (hunk 3 above)
Comment (avishkajayasundara) [reply components/org.wso2.carbon.identity.api.user.biometric.device/org.wso2.carbon.identity.api.user.biometric.device.common/src/main/java/org/wso2/carbon/identity/api/user/biometric/device/common/util/Constants.java]: Bio-metric Device Handler

**Pull request closed without merging**